pkg/operator/ceph/object/health.go

```diff
@@ -0,0 +1,272 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package object
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	s3UserHealthCheckName      = "rook-ceph-internal-s3-user-checker"
+	s3HealthCheckBucketName    = "rook-ceph-internal-s3-bucket-checker"
+	defaultStatusCheckInterval = 1 * time.Minute
+	s3HealthCheckObjectBody    = "Test Rook Object Data"
+	s3HealthCheckObjectBodyMD5 = "5f286306a2227a156ba770800e71b796"
+	s3HealthCheckObjectKey     = "rookObj1"
+	contentType                = "plain/text"
+)
+
+// bucketChecker aggregates the mon/cluster info needed to check the health of the monitors
+type bucketChecker struct {
+	context         *clusterd.Context
+	objContext      *Context
+	interval        time.Duration
+	serviceIP       string
+	port            string
+	client          client.Client
+	namespacedName  types.NamespacedName
+	healthCheckSpec *cephv1.HealthCheckSpec
+}
+
+// newbucketChecker creates a new HealthChecker object
+func newBucketChecker(context *clusterd.Context, objContext *Context, serviceIP, port string, client client.Client, namespacedName types.NamespacedName, healthCheckSpec *cephv1.HealthCheckSpec) *bucketChecker {
+	c := &bucketChecker{
+		context:         context,
+		objContext:      objContext,
+		interval:        defaultStatusCheckInterval,
+		serviceIP:       serviceIP,
+		port:            port,
+		namespacedName:  namespacedName,
+		client:          client,
+		healthCheckSpec: healthCheckSpec,
+	}
+
+	// allow overriding the check interval with an env var on the operator
+	checkInterval := healthCheckSpec.Endpoint.Interval
+	if checkInterval != "" {
+		if duration, err := time.ParseDuration(checkInterval); err == nil {
+			logger.Infof("ceph rgw status check interval for object store %q is %q", namespacedName.Name, checkInterval)
+			c.interval = duration
+		}
+	}
+
+	return c
+}
+
+// checkObjectStore periodically checks the health of the cluster
+func (c *bucketChecker) checkObjectStore(stopCh chan struct{}) {
+
+	for {
+		select {
+		case <-stopCh:
+			// purge bucket and s3 user
+			// Needed for external mode where in converged everything goes away with the CR deletion
+			c.cleanupHealthCheck()
+			logger.Infof("stopping monitoring of rgw endpoints for object store %q", c.namespacedName.Name)
+			return
+
+		case <-time.After(defaultStatusCheckInterval):
+			logger.Debug("checking rgw health of object store %q", c.namespacedName.Name)
+			err := c.checkObjectStoreHealth()
+			if err != nil {
```

Comment:
Seems like this is where the CR status should be added with the results of the health check. Then there is a single place that the status will be updated instead of in many places where `updateStatusEndpoint()` is called in failed scenarios.